rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,82 @@
+
+Certain random ops, such as `tf.image.sample_distorted_bounding_box` and `tf.nn.fractional_max_pool`, ignore the global seed if a seed is not explicitly passed. For such ops, setting the global seed is not enough to avoid the error, so users must pass a seed directly to the op.
+
+As for TensorFlow 2 random number generation, `tf.random.Generator.from_non_deterministic_state` will raise an error if called when determinism is enabled. In such cases, users should check if determinism is enabled and if so, use a different generator from a deterministic source. `tf.random.get_global_generator` implicitly calls `from_non_deterministic_state` if no global generator is set, and so will also raise an error if a global generator is not set with `tf.random.set_global_generator`.
+
+Stateless random functions, such as `tf.random.stateless_normal`, are always deterministic and so will never raise determinism-related errors.
+
+In graph mode, ops will raise an error message when the random op is created. If a random op is created in graph mode without determinism being enabled but then later runs when determinism is enabled, it will also raise an error.
+
+No error will be raised if a random op or generator is run before determinism is enabled (as is true for any other op), so users should take care to enable determinism before running any random ops or generators.
+
+### Parameter Server
+Use of parameter servers adds nondeterministic behavior. In case a model constructs a ParameterServerStrategy, TensorFlow will throw an error. We’ll also document this in the documentation for the flag.
+
+### Op Review and changes
+As part of the implementation, we will review all ops to make a determination of their behavior (deterministic vs nondeterministic). Some of the ops that are known to be nondeterministic, at least when running on a GPU, include:
+
+* `tf.nn.softmax_cross_entropy_with_logits`
+* `tf.nn.sparse_softmax_cross_entropy_with_logits`
+* `tf.image.resize` gradient with `method=ResizeMethod.NEAREST`
+* `tf.math.segment_sum`, `tf.math.unsorted_segment_sum` forward
+* `tf.image.crop_and_resize` gradient to both image and boxes 
+* `tf.sparse.sparse_dense_matmul` forward
+* `tf.math.unsorted_segment_mean`, `tf.math.unsorted_segment_prod` and `tf.math.unsorted_segment_sqrt`; all foward
+* `tf.sparse.sparse_dense_matmul`
+
+
+`tf.image.sample_distorted_bounding_box` has been observed to behave nondeterministically unless you set its seed parameter, even if you call tf.random.set_seed. We will review this Op as part the change. Another case that needs review is "pulling a random number from a PRNG before its state has been initialized".
+
+Given the large number of ops involved, there is a chance that we might omit raising an error for a nondeterministic Op.
```

Comment:
I think that this should read:

Nondeterminism introduced by ops has already been addressed in roughly priority order, add this process will continue with the new API. Even after the relatively small number of remaining ops that are known to introduce nondeterminism will throw an error after calling `tf.config.enable_deterministic_execution(True)`, other ops may be discovered to introduce nondeterminism. These will be isolated and confirmed as they arise and appropriate error throwing will be added to them.

Reply:
I don't think the first sentence is worth mentioning, as we already described that determinism is already partially implemented through environmental variables.  The second sentence is a more verbose form of what is already there. I reworded and added the third sentence: "We will fix such cases as they arises.".

Reply:
My intention is to try to convey the way I think this will actually play out, which is similar to how it has been playing out. The currently existing message is essentially that we will audit all the ops and that we might miss some. That would not be honest because it's almost certainly not going to happen because it's simply not practical. I doubt that we're going to do a determinism study on every op on every device (99% of which will always operate deterministically) and then maybe miss some because a study can only confirm that there is nondeterminism not that there is not nondeterminism.